From 997f3e28ca6bea8727edaeb54c16df41c6e1629e Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Thu, 28 Jan 2021 15:44:54 -0500
Subject: [PATCH 1/2] arm64: dts: rockchip: set type-c dr_mode as otg

This goes against what 75152d66315521a48c4997305f4e01c5f139e160 did,
where they set the rockpro64 type-c to host due to reliability issue.

The change is meant to allow using the type-c port for gadget mode.

The role switch, and role switch mode are meant to allow userspace
control of the mode, using role switch.
---
 arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
index d7fe1d99b546d..131f1d3ab579b 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
@@ -1093,7 +1093,9 @@ &usbdrd3_0 {
 };
 
 &usbdrd_dwc3_0 {
-	dr_mode = "host";
+	dr_mode = "otg";
+	usb-role-switch;
+	role-switch-default-mode = "host";
 	status = "okay";
 };
 
-- 
2.33.0


From 6923f5b2f23723c902bfcc855df2c8c155afa13a Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Thu, 28 Jan 2021 15:42:29 -0500
Subject: [PATCH 2/2] usb: dwc3: Enable userspace role switch control

Self-explanatory. Allows end-users to control the role of the controller
without having to rely on debugfs.
---
 drivers/usb/dwc3/drd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/usb/dwc3/drd.c b/drivers/usb/dwc3/drd.c
index d7f76835137fa..eb23ce88f3b6b 100644
--- a/drivers/usb/dwc3/drd.c
+++ b/drivers/usb/dwc3/drd.c
@@ -555,6 +555,7 @@ static int dwc3_setup_role_switch(struct dwc3 *dwc)
 	dwc3_role_switch.set = dwc3_usb_role_switch_set;
 	dwc3_role_switch.get = dwc3_usb_role_switch_get;
 	dwc3_role_switch.driver_data = dwc;
+	dwc3_role_switch.allow_userspace_control = true,
 	dwc->role_sw = usb_role_switch_register(dwc->dev, &dwc3_role_switch);
 	if (IS_ERR(dwc->role_sw))
 		return PTR_ERR(dwc->role_sw);
-- 
2.33.0

